pkgs/applications/editors/eclipse/plugins.nix

```diff
@@ -15,80 +34,123 @@ rec {
-
-      in
-
-        buildEclipsePluginBase (attrs // {
-          srcs = [ srcFeature ] ++ pSrcs;
-
-          buildCommand = ''
-            dropinDir="$out/eclipse/dropins/${name}"
-
-            mkdir -p $dropinDir/features
-            unzip ${srcFeature} -d $dropinDir/features/
+  # Mirror a subset of a p2 update site containing the given features and
+  # plugins.
+  mirrorEclipsePlugins = { pname
+                         , version
+                         , repository ? null
+                         , artifactRepository ? null
+                         , metadataRepository ? null
+                         , installableUnit ? null
+                         , installableUnits ? null
+                         , eclipse ? eclipses.eclipse-sdk
+                         , hash
+                         , ... } @ attrs:
+    assert (artifactRepository == null) == (metadataRepository == null);
+    assert (repository == null) != (artifactRepository == null);
+    assert (installableUnit == null) != (installableUnits == null);
+
+    let
+      ius =
+        if installableUnit != null
+        then [ installableUnit ]
+        else installableUnits;
+      iuEntries = (lib.strings.concatStrings (map (iu: "<iu id=\"${iu.id}\" version=\"${iu.version}\" />") ius));
+
+    in
+      buildEclipsePluginBase rec {
+        name = "${pname}-${version}";
+        inherit version;
+
+        buildCommand = ''
+          mkdir -p $out
+          echo '
+            <project name="${name}" default="mirror" basedir="'$out'">
+              <target name="mirror">
+                <p2.mirror raw="true">
+                  <source>
+                    ${if repository != null then "<repository location=\"${repository}\" />" else ""}
+                    ${if metadataRepository != null then "<repository location=\"${metadataRepository}\" kind=\"metadata\" />" else ""}
+                    ${if artifactRepository != null then "<repository location=\"${artifactRepository}\" kind=\"artifact\" />" else ""}
+                  </source>
+                  <destination compressed="true" name="${name}" location="file:'$out'" />
+                  ${iuEntries}
+                  <slicingOptions followStrict="true"/>
+                </p2.mirror>
+              </target>
+            </project>' > /build/build.xml
+          ${eclipse}/bin/eclipse \
+            -noSplash \
+            -configuration /build/configuration \
+            -application org.eclipse.ant.core.antRunner \
+            -buildfile /build/build.xml
+        '' + (sanitizeUpdateSite "$out");
+
+        nativeBuildInputs = [ zip eclipse xmlstarlet ];
+        outputHash = hash;
+        outputHashMode = "recursive";
+
+        passthru = {
+          installableUnits = map (iu: "${iu.id}/${iu.version}") ius;
+        };
+
+        inherit repository artifactRepository metadataRepository;
+      };
+
+  # Publish the given plugin and feature jars as a p2 update site.
```

Comment:
This shouldn't be used under normal circumstances, but it needs to be here for two reasons.
- Backwards compatibility.
- Plugins which aren't published in p2 repos. This is *extremely* rare these days in the Eclipse ecosystem!